### PR TITLE
Star Wars D6 v.2.41

### PR DESCRIPTION
**Pull request by Moondancer83**
## Changes / Comments

- refactor sheet code to use [CSE](https://wiki.roll20.net/CSE)
- implement datalists for skillnames
- various minor changes.
- fixes an issue that cause breaking attribute, skill, weapon and armour lines in zoom level <100%.
- fixes several style glitches
image

## Roll20 Requests

Comments are very helpful for reviewing the code changes. Please answer the relevant questions below in your comment.

- [x] Does the pull request title have the sheet name(s)? Include each sheet name.
- [x] Is this a bug fix?
- [x] Does this add functional enhancements (new features or extending existing features) ?
- [ ] Does this add or change functional aesthetics (such as layout or color scheme) ? 
- [ ] If changing or removing attributes, what steps have you taken, if any, to preserve player data ?
- [ ] If this is a new sheet, did you follow [Building Character Sheets standards](https://wiki.roll20.net/Building_Character_Sheets#Roll20_Character_Sheets_Repository) ?

If you do not know English. Please leave a comment in your native language.

### CODE REVIEW

**Comment by Moondancer83** [comment]
Doing some more testing.

**Comment by Kevak** [comment]
It appears as though the most recent 'update' to the WEG Star Wars D6 (At some point over the last three days) has irrevocably ruined and otherwise disrupted quite literally every PC and NPC's sheets in my WEG SWD6 game. 1's are now 0's. 2's are now 0's. 3's are now 1's. And a myriad of misc number alterations. I'd like to know whomst exactly gave you the go-ahead for this travesty of an 'update.' 'Layout Enhancement' my _ass._ 

You might have just killed the campaign I was partaking in permanently! Have you never heard of testing your code? I doubt reversion even will fix it. 

Also the new UI is ass, to add insult to injury. There's a massive amount of wasted blank space, stats and skills alone takes up thrice the space it used to. The only notable improvements I can see are the active-toggles for force powers and the name/gender/similar section being more readable.

Everything else is crap and should be reverted.

**Comment by Kevak** [comment]
Hell on that note, why the fuck didn't you make it an optional update if you were gonna change shit so drastically that you broke the entire godsdamned sheet for everyone currently playing the game?

You could have made it an optional, off by default toggle. You did not.

**Comment by Moondancer83** [comment]
Dear Kevak,

First, thank you for your honest feedback.
Secondly, I'm pretty sorry that you encountered those terrible issues. However I do not really understand them as I only modified design and layout, not behaviour or data. And of course I do testing with characters from version 2.32 without data change.
In addition all my 1+ year old campaign characters (3 PCs of real players and 10+ NPCs) are fully functional and without change.

Anyways, please help me fix the issues you encountered.

File: D6StarWars/D6StarWars.css [hunk 1]
```diff
@@ -8,11 +8,23 @@
...
+.charsheet div.version-top{
+    position: absolute;
+    right: 5px;
+    top: 30px;
+}
+
+.charsheet div.logo{
+    display: inline-block;
+    position: absolute;
+    left: 150px;
```
Comment (Kevak) [inline D6StarWars/D6StarWars.css:20]: Every single block has been manually set. This results in block overflow and will fuck larger sheets.

File: D6StarWars/D6StarWars.css [hunk 2]
```diff
@@ -337,81 +385,438 @@ input.sheet-options-flag.sheet-main-options + span {
...
+.charsheet textarea.specabilities{
+    width: 95%;
+    height: 12em;
+}
+
+.charsheet th.init{
+    vertical-align: top;
+}
+.charsheet td.init{
+    padding-left: 6px;
```
Comment (Kevak) [inline D6StarWars/D6StarWars.css:456]: As I said. Every single one. Note that this also renders the sheets incompatible with normal roll20.

File: D6StarWars/D6StarWars.css [hunk 3]
```diff
@@ -774,249 +1212,3 @@ div.sheet-section.sheet-ship div.sheet-other{
...
-}
-
-
-.charsheet .sheet-redtxt {
-    bottom: -3px;
-    right: 5px;
-}
-
-
-/* Edit Mode */
```
Comment (Kevak) [inline D6StarWars/D6StarWars.css:910]: Edit mode is broken by the way for certain blocksets.

File: D6StarWars/D6StarWars.html [hunk 4]
```diff
@@ -162,12 +161,12 @@
...
-        <div class='col'>
+        <div class="sheet-col">
           <table class="custom-roll-table">
             <tr>
-              <th style="vertical-align: top;">Initiative bonus</th>
-              <td colspan="2" style="padding-left: 6px; padding-bottom: 40px">
-                <input class="smallnumber shortnumber" min="0" name="attr_initiative" type="number" value="0"/>D+<input class="pipnumber shortnumber" max="2" min="0" name="attr_initiativepip" type="number" value="0"/>
+              <th class="init">Initiative bonus</th>
+              <td colspan="2" class="init">
+                <input class="init" min="0" name="attr_initiative" type="number" value="0"/>D+<input class="init" max="2" min="0" name="attr_initiativepip" type="number" value="0"/>
```
Comment (Kevak) [inline D6StarWars/D6StarWars.html:169]: Every time the sheet is initialized, this includes via pop-out, there is a chance that all sheet values will be reset to default value. What wonderful coding!

File: D6StarWars/D6StarWars.html [hunk 5]
```diff
@@ -983,26 +985,26 @@ <h4>Force Powers</h4>
...
-        <div style="margin-bottom:10px;">
-          <span>Damage:<input class="smallnumber" name="attr_weapon_damage" type="number" value="3">D+<input
-              class="smallnumber" name="attr_weapon_damage_pip" type="number" value="0"> <button class="d6-dice"
-                                                                                                 name="roll_ship_wpn_dmg"
-                                                                                                 type="roll"
-                                                                                                 value="&{template:swd6} {{name=Weapon: @{weapon_name} }} {{subtitle=Ship: @{ship_name} }} {{Roll=[[(@{weapon_damage}- 1)d6cf0cs0 + @{weapon_damage_pip} +(1d6@{wilddie})]] damage}}">DMG</button></span>
+        <div>
+          <span>Damage:
+            <input class="40px" name="attr_weapon_damage" type="number" value="3">D+<input
+              class="smallnumber" name="attr_weapon_damage_pip" type="number" value="0">
```
Comment (Kevak) [inline D6StarWars/D6StarWars.html:1005]: Hee hoo more value setting! This means that the value is set for **EVERY** sheet!

**Review by Kevak — COMMENTED** [review]
I think this is sufficient to point out the problems. Why did you do this?

**Comment by Kevak** [comment]
> Dear Kevak,
> 
> First, thank you for your honest feedback.
> Secondly, I'm pretty sorry that you encountered those terrible issues. However I do not really understand them as I only modified design and layout, not behaviour or data. And of course I do testing with characters from version 2.32 without data change.
> In addition all my 1+ year old campaign characters (3 PCs of real players and 10+ NPCs) are fully functional and without change.
> 
> Anyways, please help me fix the issues you encountered.

I hope this review will give you an idea of what you did wrong. I decided to not make review comments for every single duplicate fuckup. clearly your testing was inadequate or you flat didn't notice.

**Comment by Kevak** [comment]
Force-setting values to a specific number **DELETES** what the players and GM's have it set at. Let me repeat that. By forcibly setting Value=1 for a given attribute, lets say PER, every single time that sheet is initialized, it has a chance of PER being set to 1. You did not take adequate steps to preserve player data. This was not a mere graphical update, you rewrote the entire fucking attribute system and did this to it.

**Comment by Kevak** [comment]
This applies for Value=2, Value=3, and so on. Init is Initialize. You decided to stick it before Value setting. This means that every time the sheet is fucking initialized, it has a high fucking chance of setting value to the default listed value! You did this. This was not part of the original valuecode.

File: D6StarWars/D6StarWars.html (hunk 4 above)
Comment (Moondancer83) [reply D6StarWars/D6StarWars.html:169]: Sorry to say but this should not be true and I have never encountered this behaviour. Could you please send a video/gif that shows this resetting behaviour? I would really appreciate that.

EDIT:
The original sheet (v2.32 form February) has all the defaults (with _3D+0_).

EDIT:
In every frontend framework I worked with (including Roll20) the value attribute defines only a default value that WILL be overwritten any incoming external data (e.g.: database or manual input).

**Comment by NopeDotWav** [comment]
Screenshot (9935)
Hi, I made a quick test sheet after seeing that all of my sheets in the game I'm running have been altered and fucked for the most part and everything is 1D instead of 3D, are you sure your shit isn't busted? asking because this has actively destroyed my game and I really want an explanation.

File: D6StarWars/D6StarWars.html (hunk 4 above)
Comment (Kevak) [reply D6StarWars/D6StarWars.html:169]: Speak of the fucking devil. The values absolutely **ARE** fucked and we've got another person present to show your buggerup. I don't even need to waste the time making a video of a new sheet. You are wrong. You have fucked up. You did in fact, screw up. Quit trying to save face and fucking revert the PR and submit a public apology already.

Comment (Kevak) [reply D6StarWars/D6StarWars.html:169]: I'll also note you haven't disputed you've fucked the blocks sizings.

**Comment by Moondancer83** [comment]
Dear @NopeDotWav,

Thank you for your comment and feedback. I hope all your destroyed games are back after the revert happened yesterday.

Still if you encounter 3D to 1D transition this is because you have never changed the sheets default values to real user inputs. Basically there is no data loss as there was no data saved at all.

I'm not saying this is your fault. **It is mine.** I assumed everybody changes default values and my assumption was wrong.

However, having no persisted values for a skill for example can cause issues in normal roll mechanism as well even without my changes. Reasoning: the roll mechanism uses the persisted values that can be checked at **Attributes & Abilities** tab.

If you have encountered any other kind of data loss related to my prior change please notify me.


Once more you are right, I did test all the cases regarding the PC sheet change that I was aware of.
I missed testing all the possible browser zoom levels that caused some fractions in the layout. These issues are fixed for now.
I missed testing the other sheet types (sheet, container) as I tried to change only PC (NPC) related code. I was wrong, again, with the scoping.

Thanks

**Comment by Moondancer83** [comment]
Dear @Kevak,
> This applies for Value=2, Value=3, and so on. Init is Initialize. You decided to stick it before Value setting. This means that every time the sheet is fucking initialized, it has a high fucking chance of setting value to the default listed value! You did this. This was not part of the original valuecode.
Your above statement is simply not true.

Other than this I accept all the blame about not testing long & wide enough and making false assumptions.

I would be happy if you could help me fixing the issues we have as I see much energy and willingness of helping in you. Just please be a bit less harsh with others even if they are fucking dumb-asses like me.

Thank you.

**Pull request closed without merging**